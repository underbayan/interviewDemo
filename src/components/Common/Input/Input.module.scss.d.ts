export const inputStyle: string;
export const spanStyle: string;
export const containerStyle: string;
