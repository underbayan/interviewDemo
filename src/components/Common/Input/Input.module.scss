@import "@demo/designSystem";

.containerStyle {
  display: flex;
  align-items: baseline;

  .inputStyle {
    padding: 4px 8px;
    outline: none;
    border: none;
    border-bottom: var(--grays-5-color) 1px solid;
    color: var(--grays-3-color);

    &::placeholder {
      color: nth($grays, 4);
    }

    &:focus {
      border-bottom: 1px solid var(--grays-3-color);
    }
  }

  .spanStyle {
    align-self: flex-end;
    margin-inline-start: 12px;
    font-size: 0.8em;
  }
}