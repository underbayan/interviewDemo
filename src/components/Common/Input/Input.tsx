import {
  forwardRef,
  useState,
  useImperativeHandle,
  memo,
  RefObject,
  ChangeEventHandler,
  ChangeEvent,
  InputHTMLAttributes,
  useCallback,
} from 'react';
import { inputStyle, spanStyle, containerStyle } from './Input.module.scss';
import { omit } from 'lodash';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  className?: string;
  onChange?: (value: ChangeEvent<HTMLInputElement>) => void;
  message?: string;
  validate?: (value: string) => boolean;
}

export interface InputRef {
  ok: boolean;
  value: string;
  setOk: (ok: boolean) => void;
  setValue: (value: string) => void;
}

const Input = memo(
  forwardRef<InputRef, InputProps>((props, ref) => {
    const [value, setValue] = useState<string>('');
    const [flag, setFlag] = useState<string>('');
    const [ok, setOk] = useState<boolean>(false);
    const { className, onChange, message, validate, ...rest } = props;
    console.log(121212)
    useImperativeHandle<InputRef, InputRef>(
      ref as RefObject<InputRef>,
      () => ({ ok, value, setOk, setValue }),
      [ok, value, setOk, setValue],
    );

    const handleChange: ChangeEventHandler<HTMLInputElement> = useCallback(e => {
      setValue(e.target.value);
      const isOk = validate ? validate(e.target.value) : true;
      setOk(isOk);
      const newFlag = !e.target.value || !validate ? '' : isOk ? '✔️' : '❌';
      if (newFlag !== flag) setFlag(newFlag);
      onChange && onChange(e.target.value as unknown as ChangeEvent<HTMLInputElement>);
    }, [onChange, setFlag, flag, validate, setValue]);

    return (
      <div className={`${containerStyle} p-t-xs-1`}>
        <input
          type="text" // should be override
          autoComplete="on"
          {...omit(rest, 'validate')}
          className={`${inputStyle} ${className ?? ''}`}
          value={value}
          ref={ref as RefObject<HTMLInputElement>}
          onChange={handleChange}
        />
        {value ? <span className={spanStyle}>{flag}</span> : null}
        {message ? <span className={spanStyle}>{message}</span> : null}
      </div>
    );
  }),
);

export { Input };
