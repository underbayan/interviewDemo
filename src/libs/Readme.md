"Libs" refer to the libraries or modules that contain code that is meant to be reusable across different parts of an application. In the context of the specified constraint, the code in "libs" should not contain any mutations logic. Rather, it should be focused on providing functionalities that are shared across different parts of the application without altering any data. For instance, "libs" could contain helper functions, utility classes, or any other non-mutating code that can be used in different parts of the application.