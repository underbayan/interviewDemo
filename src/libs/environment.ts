import stringsEn from "../mock.strings.en.json"
import stringsZh from "../mock.strings.zh.json"

export const environment = {
  app: "demo",
  page: "/",
  lang: "en",
  version: "0.0.1",
  flights: "", // AB test
  config: {},  // Config for page
  strings: stringsEn,
  langMap: {
    en: stringsEn,
    zh: stringsZh
  }
  // TODO: i18n.detectLanguage() to get zh/en/fr lang config json
}