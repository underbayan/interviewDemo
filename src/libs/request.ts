export interface FetchDataType {
  [key: string]: any
}

export interface AbortableFetchResponse extends Response {
  data?: FetchDataType | string | ArrayBuffer | Blob
  message: string
}

export interface AbortableFetchOptions extends RequestInit {
  abort?: AbortSignal
}

export interface AbortableFetchPromise extends Promise<AbortableFetchResponse> {
  [AbortableKey](): void
}

export const AbortableKey = Symbol.for('abortableKey')
export const abortableFetch = (
  url: string,
  options?: AbortableFetchOptions
): AbortableFetchPromise => {
  const { abort, ...fetchOptions } = options ?? {}
  const ab = new AbortController()
  const rt = (fetch(url, { ...fetchOptions, signal: abort ? abort : ab.signal }) as Promise<Response>)
    .then(async (response) => {
      const contentType = response.headers.get('content-type')
      let data: FetchDataType | undefined | string

      if (contentType?.includes('application/json')) {
        data = await response.json()
      } else if (contentType?.includes('application/text')) {
        data = await response.text()
      } else if (contentType?.includes('application/opaque')) {
        data = await response.blob()
      } else {
        data = await response.arrayBuffer()
      }

      const result: AbortableFetchResponse = {
        ...response,
        status: response.status,
        message: response.statusText,
        data,
      }
      return result
    })
    .catch((e) => console.error(e)) as AbortableFetchPromise

  Object.defineProperty(rt, AbortableKey, { value: () => ab.abort(), writable: false })

  return rt
}
