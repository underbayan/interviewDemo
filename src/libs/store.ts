import { createSlice, configureStore } from '@reduxjs/toolkit'
import { useSelector, useDispatch } from 'react-redux'
import { environment } from "./environment"
import { Dispatch, AnyAction } from 'redux'
export type RootState = ReturnType<typeof store.getState>

const envSlice = createSlice({
  name: 'env',
  initialState: environment,
  reducers: {

    updateEnv(state, action) {

      const newState = { ...state, ...action?.payload }
      if (action?.payload?.lang) newState.strings = newState.langMap[action.payload?.lang]
      return newState
    }
  }
})
export const store = configureStore({
  reducer: {
    env: envSlice.reducer
  },
})
interface EnvUpdatePayload {
  lang?: string
  [key: string]: any
}




export const useI18n = () => useSelector((state: RootState) => state.env.strings)
export const useEnv = () => useSelector((state: RootState) => state.env)
export const useUpdateEnv = (): ((env: EnvUpdatePayload) => void) => {
  const dispatch = useDispatch()

  const updateEnvAsync = (env: EnvUpdatePayload) => (dispatch: Dispatch) => {
    setTimeout(() => { dispatch(envSlice.actions.updateEnv(env)) }, 15)
  }

  return (env: EnvUpdatePayload) => dispatch(updateEnvAsync(env) as unknown as AnyAction)
}


