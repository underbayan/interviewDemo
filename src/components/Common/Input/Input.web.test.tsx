import { render } from '@testing-library/react';
import { test, expect } from 'vitest';
import { Input } from './Input';
import { store } from "@demo/libs/store"
import { Provider } from 'react-redux'

test('Input renders correctly', () => {
  const { container } = render(<Provider store={store}>
    <Input></Input>
  </Provider>);
  expect(container).toMatchSnapshot();
});
