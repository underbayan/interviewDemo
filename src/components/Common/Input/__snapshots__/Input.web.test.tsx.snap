// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input renders correctly 1`] = `
<div>
  <div
    class="undefined p-t-xs-1"
  >
    <input
      autocomplete="on"
      class="undefined "
      type="text"
      value=""
    />
  </div>
</div>
`;
